File: simulator/backend/routes/industry-settings.ts
import express from 'express';
import { industryService } from '../services/industry-service.ts';

const router = express.Router();

// Middleware to check authentication from Bearer token
async function requireAuth(req: express.Request, res: express.Response, next: express.NextFunction) {
  try {
    const token = req.headers.authorization?.replace('Bearer ', '');
    if (!token) {
      return res.status(401).json({ success: false, error: 'Authentication required' });
    }

    const { sessionRepository } = await import('../db/repositories/session-repository.ts');
    const session = await sessionRepository.findByToken(token);

    if (!session) {
      return res.status(401).json({ success: false, error: 'Invalid or expired session' });
    }

    const { userService } = await import('../services/user-service.ts');
    const user = await userService.getUserById(session.user_id);
    if (!user) {
      return res.status(401).json({ success: false, error: 'User not found' });
    }

    (req as any).user = user;
    next();
  } catch (error) {
    console.error('Auth error:', error);
    return res.status(401).json({ success: false, error: 'Authentication failed' });
  }
}

// Middleware to check admin role
function requireAdmin(req: express.Request, res: express.Response, next: express.NextFunction) {
  const user = (req as any).user;
  if (!user || (user.role !== 'super_admin' && user.role !== 'company_admin')) {
    return res.status(403).json({ success: false, error: 'Admin access required' });
  }
  next();
}

// GET /api/industry-settings - Get all industry settings
router.get('/', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const industryCompetencies = await industryService.getIndustryCompetencies();
    const industryMetadata = await industryService.getIndustryMetadata();
    const difficultySettings = await industryService.getDifficultySettings();

    return res.json({
      success: true,
      data: {
        industryCompetencies,
        industryMetadata,
        difficultySettings
      }
    });
  } catch (error) {
    console.error('Get industry settings error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to get industry settings'
    });
  }
});

// GET /api/industry-settings/competencies - Get industry competencies mappings
router.get('/competencies', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const industryCompetencies = await industryService.getIndustryCompetencies();

    return res.json({
      success: true,
      data: industryCompetencies
    });
  } catch (error) {
    console.error('Get industry competencies error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to get industry competencies'
    });
  }
});

// GET /api/industry-settings/metadata - Get industry metadata
router.get('/metadata', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const industryMetadata = await industryService.getIndustryMetadata();

    return res.json({
      success: true,
      data: industryMetadata
    });
  } catch (error) {
    console.error('Get industry metadata error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to get industry metadata'
    });
  }
});

// GET /api/industry-settings/difficulty - Get difficulty settings
router.get('/difficulty', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const difficultySettings = await industryService.getDifficultySettings();

    return res.json({
      success: true,
      data: difficultySettings
    });
  } catch (error) {
    console.error('Get difficulty settings error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to get difficulty settings'
    });
  }
});

// PUT /api/industry-settings/competencies/:industry/:subcategory - Update industry/subcategory competencies
router.put('/competencies/:industry/:subcategory', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const { industry, subcategory } = req.params;
    const { competencyIds } = req.body;

    if (!Array.isArray(competencyIds)) {
      return res.status(400).json({
        success: false,
        error: 'competencyIds must be an array'
      });
    }

    const success = await industryService.updateIndustrySubcategoryCompetencies(
      industry,
      subcategory,
      competencyIds
    );

    if (!success) {
      return res.status(500).json({
        success: false,
        error: 'Failed to update industry subcategory competencies'
      });
    }

    return res.json({ success: true, message: 'Industry subcategory competencies updated successfully' });
  } catch (error) {
    console.error('Update industry subcategory competencies error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to update industry subcategory competencies'
    });
  }
});

// PUT /api/industry-settings/focus-area/:industry/:subcategory/:focusArea - Update focus area competencies
router.put('/focus-area/:industry/:subcategory/:focusArea', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const { industry, subcategory, focusArea } = req.params;
    const { competencyIds, enabled } = req.body;

    if (!Array.isArray(competencyIds)) {
      return res.status(400).json({
        success: false,
        error: 'competencyIds must be an array'
      });
    }

    const success = await industryService.updateFocusAreaCompetencies(
      industry,
      subcategory,
      focusArea,
      competencyIds,
      enabled !== undefined ? enabled : true
    );

    if (!success) {
      return res.status(500).json({
        success: false,
        error: 'Failed to update focus area competencies'
      });
    }

    return res.json({ success: true, message: 'Focus area competencies updated successfully' });
  } catch (error) {
    console.error('Update focus area competencies error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to update focus area competencies'
    });
  }
});

// PUT /api/industry-settings/difficulty/:industry - Save difficulty settings for industry
router.put('/difficulty/:industry', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const { industry } = req.params;
    const settings = req.body;

    if (!settings || typeof settings !== 'object') {
      return res.status(400).json({
        success: false,
        error: 'Invalid difficulty settings format'
      });
    }

    const success = await industryService.saveDifficultySettings(industry, settings);

    if (!success) {
      return res.status(500).json({
        success: false,
        error: 'Failed to save difficulty settings'
      });
    }

    return res.json({ success: true, message: 'Difficulty settings saved successfully' });
  } catch (error) {
    console.error('Save difficulty settings error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to save difficulty settings'
    });
  }
});

// PUT /api/industry-settings/metadata - Save full industry metadata
router.put('/metadata', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const data = req.body;

    if (!data || typeof data !== 'object') {
      return res.status(400).json({
        success: false,
        error: 'Invalid metadata format'
      });
    }

    const success = await industryService.saveIndustryMetadata(data);

    if (!success) {
      return res.status(500).json({
        success: false,
        error: 'Failed to save industry metadata'
      });
    }

    return res.json({ success: true, message: 'Industry metadata saved successfully' });
  } catch (error) {
    console.error('Save industry metadata error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to save industry metadata'
    });
  }
});

// PUT /api/industry-settings/competencies - Save full industry competencies
router.put('/competencies', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const data = req.body;

    if (!data || typeof data !== 'object') {
      return res.status(400).json({
        success: false,
        error: 'Invalid competencies format'
      });
    }

    const success = await industryService.saveIndustryCompetencies(data);

    if (!success) {
      return res.status(500).json({
        success: false,
        error: 'Failed to save industry competencies'
      });
    }

    return res.json({ success: true, message: 'Industry competencies saved successfully' });
  } catch (error) {
    console.error('Save industry competencies error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to save industry competencies'
    });
  }
});

export default router;

File: backend-python/app/routers/industry_settings.py
"""
Industry Settings Router
Industry configuration endpoints
"""

import logging
from typing import Annotated, Any

from fastapi import APIRouter, HTTPException, status, Depends, Body

from app.services.industry_service import industry_service
from app.middleware.auth import get_current_user, require_admin
from app.models.user import UserData

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/industry-settings", tags=["Industry Settings"])


@router.get("")
async def get_industry_settings(
    user: Annotated[UserData, Depends(get_current_user)],
):
    """Get all industry settings (read access for all authenticated users)."""
    try:
        industry_competencies = await industry_service.get_industry_competencies()
        industry_metadata = await industry_service.get_industry_metadata()
        difficulty_settings = await industry_service.get_difficulty_settings()

        return {
            "success": True,
            "data": {
                "industryCompetencies": industry_competencies,
                "industryMetadata": industry_metadata,
                "difficultySettings": difficulty_settings,
            },
        }
    except Exception as e:
        logger.error(f"Get industry settings error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/competencies")
async def get_industry_competencies(
    user: Annotated[UserData, Depends(get_current_user)],
):
    """Get industry competencies mappings (read access for all authenticated users)."""
    try:
        industry_competencies = await industry_service.get_industry_competencies()

        return {"success": True, "data": industry_competencies}
    except Exception as e:
        logger.error(f"Get industry competencies error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/metadata")
async def get_industry_metadata(
    user: Annotated[UserData, Depends(get_current_user)],
):
    """Get industry metadata (read access for all authenticated users)."""
    try:
        industry_metadata = await industry_service.get_industry_metadata()

        return {"success": True, "data": industry_metadata}
    except Exception as e:
        logger.error(f"Get industry metadata error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/difficulty")
async def get_difficulty_settings(
    user: Annotated[UserData, Depends(get_current_user)],
):
    """Get difficulty settings (read access for all authenticated users)."""
    try:
        difficulty_settings = await industry_service.get_difficulty_settings()

        return {"success": True, "data": difficulty_settings}
    except Exception as e:
        logger.error(f"Get difficulty settings error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.put("/competencies/{industry}/{subcategory}")
async def update_industry_subcategory_competencies(
    industry: str,
    subcategory: str,
    admin: Annotated[UserData, Depends(require_admin)],
    competencyIds: list[str] = Body(...),
):
    """Update industry/subcategory competencies."""
    try:
        if not isinstance(competencyIds, list):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="competencyIds must be an array",
            )

        success = await industry_service.update_industry_subcategory_competencies(
            industry,
            subcategory,
            competencyIds,
        )

        if not success:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to update industry subcategory competencies",
            )

        return {"success": True, "message": "Industry subcategory competencies updated successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Update industry subcategory competencies error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.put("/focus-area/{industry}/{subcategory}/{focus_area}")
async def update_focus_area_competencies(
    industry: str,
    subcategory: str,
    focus_area: str,
    admin: Annotated[UserData, Depends(require_admin)],
    competencyIds: list[str] = Body(...),
    enabled: bool | None = Body(None),
):
    """Update focus area competencies."""
    try:
        if not isinstance(competencyIds, list):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="competencyIds must be an array",
            )

        success = await industry_service.update_focus_area_competencies(
            industry,
            subcategory,
            focus_area,
            competencyIds,
            enabled if enabled is not None else True,
        )

        if not success:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to update focus area competencies",
            )

        return {"success": True, "message": "Focus area competencies updated successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Update focus area competencies error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.put("/difficulty/{industry}")
async def save_difficulty_settings(
    industry: str,
    admin: Annotated[UserData, Depends(require_admin)],
    settings: dict[str, Any] = Body(...),
):
    """Save difficulty settings for industry."""
    try:
        if not settings or not isinstance(settings, dict):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid difficulty settings format",
            )

        success = await industry_service.save_difficulty_settings(industry, settings)

        if not success:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to save difficulty settings",
            )

        return {"success": True, "message": "Difficulty settings saved successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Save difficulty settings error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.put("/metadata")
async def save_industry_metadata(
    admin: Annotated[UserData, Depends(require_admin)],
    data: dict[str, Any] = Body(...),
):
    """Save full industry metadata."""
    try:
        if not data or not isinstance(data, dict):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid metadata format",
            )

        success = await industry_service.save_industry_metadata(data)

        if not success:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to save industry metadata",
            )

        return {"success": True, "message": "Industry metadata saved successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Save industry metadata error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.put("/competencies")
async def save_industry_competencies(
    admin: Annotated[UserData, Depends(require_admin)],
    data: dict[str, Any] = Body(...),
):
    """Save full industry competencies."""
    try:
        if not data or not isinstance(data, dict):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid competencies format",
            )

        success = await industry_service.save_industry_competencies(data)

        if not success:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to save industry competencies",
            )

        return {"success": True, "message": "Industry competencies saved successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Save industry competencies error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))
